network/src/testutils/test_node.rs

```diff
@@ -0,0 +1,313 @@
+        network_id: NetworkId,
+        conn_metadata: ConnectionMetadata,
+    ) {
+        // PeerManager pushes this data before it's received by events
+        self.peer_metadata_storage.remove(&PeerNetworkId::new(
+            network_id,
+            conn_metadata.remote_peer_id,
+        ));
+        self.connection_update_sender
+            .push(
+                conn_metadata.remote_peer_id,
+                ConnectionNotification::LostPeer(
+                    conn_metadata,
+                    // TODO will `RoleType` matter
+                    NetworkContext::new(RoleType::Validator, network_id, self_peer_id),
+                    DisconnectReason::ConnectionLost,
+                ),
+            )
+            .unwrap();
+    }
+}
+
+/// An application specific network handle
+pub type ApplicationNetworkHandle<Sender, Events> = (NetworkId, Sender, Events);
+
+/// A unique identifier of a node across the entire network
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub struct NodeId {
+    pub owner: u32,
+    pub node_type: NodeType,
+}
+
+impl NodeId {
+    pub fn validator(owner: u32) -> NodeId {
+        NodeId {
+            owner,
+            node_type: NodeType::Validator,
+        }
+    }
+
+    pub fn vfn(owner: u32) -> NodeId {
+        NodeId {
+            owner,
+            node_type: NodeType::ValidatorFullNode,
+        }
+    }
+
+    pub fn pfn(owner: u32) -> NodeId {
+        NodeId {
+            owner,
+            node_type: NodeType::PublicFullNode,
+        }
+    }
+}
+
+impl std::fmt::Display for NodeId {
+    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
+        write!(f, "{}-{:?}", self.owner, self.node_type)
+    }
+}
+
+/// An enum defining the type of node
+#[derive(Clone, Copy, Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
+pub enum NodeType {
+    Validator,
+    ValidatorFullNode,
+    PublicFullNode,
+}
+
+/// A trait defining an application specific node with networking abstracted
+///
+/// This is built as an abstract implementation of networking around a node
+pub trait TestNode {
+    fn node_id(&self) -> NodeId;
+
+    fn node_type(&self) -> NodeType;
+
+    /// For sending to this node.  Generally should not be used after setup
+    fn node_inbound_handle(&self, network_id: NetworkId) -> InboundNetworkHandle;
+
+    fn add_other_inbound_handle(
+        &mut self,
+        peer_network_id: PeerNetworkId,
+        handle: InboundNetworkHandle,
+    );
+
+    /// For sending to other nodes
+    fn other_inbound_handle(&self, peer_network_id: PeerNetworkId) -> InboundNetworkHandle;
+
+    /// For receiving messages from other nodes
+    fn outbound_handle(&mut self, network_id: NetworkId) -> &mut OutboundMessageReceiver;
+
+    fn peer_metadata_storage(&self) -> &PeerMetadataStorage;
+
+    fn peer_network_ids(&self) -> &HashMap<NetworkId, PeerNetworkId>;
+
+    fn peer_network_id(&self, network_id: NetworkId) -> PeerNetworkId {
```

Comment:
nit: if we're planning on exposing the method to get all ids in the map, why not make this a helper that gets them all and then finds it? Why both?